### ISSUE TITLE
Modularize OSA component

### ISSUE THREAD

**Issue by dtcallcock** [OxfordIonTrapGroup/wand#26]
Right now we don't have any OSAs.  When we do get them we plan on using a Stabilizer board instead of an NI card to scan them and acquire the PD signal. 

By breaking out the OSA stuff into its own module (or whatever) we would first be able to install and run unmodified WAND without all the NI dependencies.  Then when the time comes we'd be able to write a new OSA module for the stabilizer. 

I haven't looked at the code carefully so I'll leave it to @aquinn2 and @Jemetzner to suggest an implementation.

At some point it may be worth doing for the wavemeter too.  Some groups use the Bristol ones for example.  There is also the issue that High Finesse have just randomly changed the API over the years without documenting so code is only works for some range of serial numbers (though that could be handled by polling the wavemeter for it's model and serial number).

**Comment by hartytp** [comment]
Sure I get all of that. The osa stuff is already in its own module for precisely this purpose so it will be easy to add a new osa driver later on.

However you’re hitting an error we don’t see and which I don’t see an obvious cause for in our code. I’m not sure I’ll be able to help you until you run the same code as us so I can reproduce it.

Simulation mode is there for precisely this purpose as it allows you to run without hardware. Can you just install pydaqmx and try running in sim mode first? It’s a small package so shouldn’t be a problem for you to install (you may even be able to get away with commenting out the import line as your only change if you really don’t want to install it).

**Comment by hartytp** [comment]
Sorry miss read as being part of a separate issue. Anyway afaict this is already done. The osa/wavemeter drivers are in their own models already as this was anticipated from the outset. Once someone writes a driver for a new osa/wavemeter I’ll gladly help them integrate it into wand.

**Comment by dtcallcock** [comment]
> The osa stuff is already in its own module for precisely this purpose so it will be easy to add a new osa driver later on.

Indeed, I see that now.  

> Simulation mode is there for precisely this purpose as it allows you to run without hardware.

The problem is that we want simulation mode for the OSAs only.  I think this is why the students hacked the code up a bunch.  Looking more carefully at the code it looks like it wouldn't be too hard to implement something where you can specify which bits are in simulation mode, but I'll have to chat with the students on Mon to see why they didn't go that route.  A 'no OSA' OSA model might make more sense in the longer run than having to put some random OSA model into simulation mode.

> until you run the same code as us so I can reproduce it.

Yes, I'll get them to do it on Monday.

**Comment by hartytp** [comment]
> A 'no OSA' OSA model might make more sense in the longer run than having to put some random OSA model into simulation mode.

I had a think about this, but it's a little messy. Do we add three different command-line arguments: `simulate-osa`, `simulate-wlm` and `simulate-switch`? Or, we could add drivers for "dummy" wlm, osa and switch.

All doable and not that much work, but is it really worth it? Simulation mode is intended for quick testing, I don't anticipate anyone wanting to operate in simulation mode for extended periods of time (you're in a pretty unusual situation, being blocked by waiting for new hardware to be developed). So, the complexity/time trade-off and potential to have bit-rotten code once you get your osas up and running seems quite high.

For testing, all you need to do is change this line https://github.com/OxfordIonTrapGroup/wand/blob/2bbfebe8b0a08f370a7e700408ef5fe1e20e7016/wand/frontend/wand_server.py#L92 to put the osa in simulation mode permanently.

In any case, if you want to discuss this any further, it's worth opening up a new issue.

**Comment by hartytp** [comment]
> A 'no OSA' OSA model might make more sense in the longer run than having to put some random OSA model into simulation mode.

Do you mean a dummy OSA driver that can be loaded instead of running the OSA in simulation mode? If so then I'd be fine with that approach. We could even have a config option that determines what kind of data the dummy OSA broadcasts (no data/simulated data/etc). Again, that would be a separate issue however.
